src/AvroConvert/AvroObjectServices/Write/Resolvers/Union.cs

```diff
@@ -56,19 +57,19 @@ private bool UnionBranchMatches(TypeSchema typeSchema, object obj)
                 case AvroType.Null:
                     return obj == null;
                 case AvroType.Boolean:
-                    return obj is bool;
+                    return obj is bool || typeof(bool).CanChangeTypeFrom(obj);
                 case AvroType.Int:
-                    return (obj is short or ushort or int or uint or char or byte or sbyte);
+                    return (obj is short or ushort or int or uint or char or byte or sbyte) || typeof(int).CanChangeTypeFrom(obj);
                 case AvroType.Long:
-                    return (obj is long or ulong);
+                    return (obj is long or ulong) || typeof(long).CanChangeTypeFrom(obj);
                 case AvroType.Float:
-                    return obj is float;
+                    return obj is float || typeof(float).CanChangeTypeFrom(obj);
                 case AvroType.Double:
-                    return obj is double;
+                    return obj is double || typeof(double).CanChangeTypeFrom(obj);
```

Comment:
These checks are needed for allowing conversions between declared property types, and actual values. Eg. a property type is an int but serialized/deserialized as a double (see unit tests).